docker/main/ngen/py_funcs.py

```diff
@@ -0,0 +1,310 @@
+
+
+def _parse_args() -> argparse.Namespace:
+    """
+    Set up and run top-level arg parsing for module.
+
+    Returns
+    -------
+    argparse.Namespace
+        The parsed arguments namespace object.
+    """
+    parser = argparse.ArgumentParser(formatter_class=argparse.ArgumentDefaultsHelpFormatter, prog='py_funcs',
+                                     description="Run one of several Docker image entrypoint Python helper functions.")
+    parser.add_argument('--log-level', '-L', dest='log_level',
+                        default=os.environ.get("DEFAULT_LOG_LEVEL", "INFO").upper(),
+                        choices=["CRITICAL", "ERROR", "WARNING", "INFO", "DEBUG"], help='Optionally specify log level.')
+
+    subparsers = parser.add_subparsers(dest='command', help='Specify the Python helper function command to run.')
+    subparsers.required = True
+
+    _parse_for_tar_and_copy(parent_subparsers_container=subparsers)
+    _parse_for_gather_output(parent_subparsers_container=subparsers)
+    _parse_for_move_job_output(parent_subparsers_container=subparsers)
+
+    return parser.parse_args()
+
+
+def _move_to_directory(source_dir: Path, dest_dir: Path, archive_name: Optional[str] = None):
+    """
+    Move source data files from their initial directory to a different directory, potentially combining into an archive.
+
+    Parameters
+    ----------
+    source_dir
+        The source directory containing output files to be placed within the dataset; note that this has already been
+        tested for existence as a directory.
+    dest_dir
+        The destination directory, to which the output data should be moved; note that this path has not yet been tested
+        for existence as a directory.
+    archive_name
+        When files should be archived as part of moving, the name of the archive; when files should not be archived,
+        ``None``.
+    """
+    if not dest_dir.is_dir():
+        raise ValueError(f"{get_date_str()} Can't move job output to non-directory path {dest_dir!s}!")
+
+    if archive_name:
+        logging.info("Archiving output files to output dataset")
+        tar_and_copy(source=source_dir, dest=dest_dir, archive_name=archive_name)
+    else:
+        logging.info("Moving output file(s) to output dataset")
+        for p in source_dir.glob("*"):
+            shutil.move(p, dest_dir)
+
+
+def gather_output(mpi_host_names: List[str], output_write_dir: Path):
+    """
+    Using subprocesses, gather output from remote MPI hosts and collect in the analogous directory on this host.
+
+    Parameters
+    ----------
+    mpi_host_names
+        List of MPI host names for the relevant job.
+    output_write_dir
+        Common job output directory across all hosts, from which data on remotes should be taken and in which data on
+        this host should be collected.
+    """
+    from socket import gethostname
+    local_hostname = gethostname()
+
+    scp_subs = OrderedDict()
```

Comment:
Why is an `OrderedDict` used here? as of 3.7, `dict` and `OrderedDict` are nearly identical. The only major differences are how `OrderedDict`s are considered equal to other `OrderedDict`s (but not other maps), a few extra operators used for merging, and different behavior for `popitem` and `reversed`.

Reply:
Gotcha.  I've taken it out.